src/System.IO/tests/Stream/Stream.CopyToTests.cs

```diff
@@ -0,0 +1,338 @@
+            // in a negative bufferSize.
+
+            var baseStream = new DelegateStream(
+                canReadFunc: () => true,
+                canSeekFunc: () => true,
+                lengthFunc: () => length,
+                positionGetFunc: () => position,
+                readFunc: (buffer, offset, count) => 0);
+            var trackingStream = new CallTrackingStream(baseStream);
+
+            var dest = Stream.Null;
+            trackingStream.CopyTo(dest);
+
+            // We don't need to do any argument checking here;
+            // if bufferSize was null CopyTo will throw an exception
+            // during argument validation
+        }
+
+        [Theory]
+        [MemberData(nameof(LengthMinusPositionPositiveOverflows))]
+        public async void AsyncIfLengthMinusPositionPositiveOverflowsBufferSizeShouldStillBePositive(long length, long position)
+        {
+            var baseStream = new DelegateStream(
+                canReadFunc: () => true,
+                canSeekFunc: () => true,
+                lengthFunc: () => length,
+                positionGetFunc: () => position,
+                readFunc: (buffer, offset, count) => 0);
+            var trackingStream = new CallTrackingStream(baseStream);
+
+            var dest = Stream.Null;
+            await trackingStream.CopyToAsync(dest);
+            
+            // We do need to check the arguments here, since both
+            // of the mock streams override CopyToAsync and don't
+            // validate the arguments
+
+            Assert.Same(dest, trackingStream.CopyToAsyncDestination);
+            Assert.True(trackingStream.CopyToAsyncBufferSize > 0);
+            Assert.Equal(CancellationToken.None, trackingStream.CopyToAsyncCancellationToken);
+        }
+
+        [Theory]
+        [MemberData(nameof(LengthIsGreaterThanPositionAndDoesNotOverflow))]
+        public void IfLengthIsGreaterThanPositionAndDoesNotOverflowEverythingShouldGoNormally(long length, long position)
+        {
+            const int ReadLimit = 7;
+
+            // Lambda state
+            byte[] outerBuffer = null;
+            int? outerOffset = null;
+            int? outerCount = null;
+            int readsLeft = ReadLimit;
+
+            var srcBase = new DelegateStream(
+                canReadFunc: () => true,
+                canSeekFunc: () => true,
+                lengthFunc: () => length,
+                positionGetFunc: () => position,
+                readFunc: (buffer, offset, count) =>
+                {
+                    Assert.NotNull(buffer);
+                    Assert.True(offset >= 0 && offset + count <= buffer.Length);
+                    Assert.True(count > 0);
+
+                    // CopyTo should always pass in the same buffer/offset/count
+                    
+                    if (outerBuffer != null) Assert.Same(outerBuffer, buffer);
+                    else outerBuffer = buffer;
```

Comment:
IMO I like this version better since it makes is clearer that we're testing against `buffer` rather than `outerBuffer`; I'm deciding to keep it the way it is.